l2tol2msg/indexer.go

```diff
@@ -0,0 +1,183 @@
+package l2tol2msg
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/asaskevich/EventBus"
+	"github.com/ethereum-optimism/optimism/op-service/predeploys"
+	"github.com/ethereum-optimism/optimism/op-service/tasks"
+	"github.com/ethereum-optimism/supersim/bindings"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+type L2ToL2MessageIndexer struct {
+	storeManager *L2ToL2MessageStoreManager
+	eb           EventBus.Bus
+	clients      map[uint64]*ethclient.Client
+	tasks        tasks.Group
+	tasksCtx     context.Context
+	tasksCancel  context.CancelFunc
+}
+
+func NewL2ToL2MessageIndexer() *L2ToL2MessageIndexer {
+	tasksCtx, tasksCancel := context.WithCancel(context.Background())
+
+	return &L2ToL2MessageIndexer{
+		storeManager: NewL2ToL2MessageStoreManager(),
+		eb:           EventBus.New(),
+		tasks: tasks.Group{
+			HandleCrit: func(err error) {
+				fmt.Printf("unhandled indexer error: %v\n", err)
+			},
+		},
+		tasksCtx:    tasksCtx,
+		tasksCancel: tasksCancel,
+	}
+}
+
+func (i *L2ToL2MessageIndexer) Start(ctx context.Context, clients map[uint64]*ethclient.Client) error {
+	i.clients = clients
+
+	for chainID, client := range i.clients {
+		client := client
+
+		i.tasks.Go(func() error {
+
+			logCh := make(chan types.Log)
+
+			sub, err := client.SubscribeFilterLogs(i.tasksCtx, ethereum.FilterQuery{
+				Addresses: []common.Address{predeploys.L2toL2CrossDomainMessengerAddr},
+			}, logCh)
+			if err != nil {
+				return fmt.Errorf("failed to subscribe to L2ToL2CrossDomainMessenger events: %w", err)
+			}
+
+			for {
+				select {
+				case log := <-logCh:
+					if err := i.processEventLog(i.tasksCtx, client, chainID, &log); err != nil {
+						fmt.Printf("failed to process log: %v\n", err)
+					}
+				case <-i.tasksCtx.Done():
+					sub.Unsubscribe()
+				}
+			}
+		})
+	}
+	return nil
+}
+
+func (i *L2ToL2MessageIndexer) Stop(ctx context.Context) error {
+	i.tasksCancel()
+	return nil
+}
+
+func sentMessageForSourceKey(sourceChainID uint64) string {
+	return fmt.Sprintf("SentMessage:source:%d", sourceChainID)
+}
+
+func (i *L2ToL2MessageIndexer) SubscribeSentMessageForSource(sourceChainID uint64, sentMessageChan chan<- *L2ToL2MessageStoreEntry) (func(), error) {
```

Comment:
Nit on the naming to make this and destination clearer
```suggestion
func (i *L2ToL2MessageIndexer) SubscribeSentMessageFromSource(sourceChainID uint64, sentMessageChan chan<- *L2ToL2MessageStoreEntry) (func(), error) {
```